design/open/spark-outputcommitter/committer.md

```diff
@@ -0,0 +1,194 @@
+auxiliary systems such as a small HDFS or S3Guard.  The magic committer
+uploads a directory tree by starting a multipart upload to its intended location,
+creating a completion record for that multipart upload on S3, and uploading
+the object without completing the upload.  To complete the upload, all
+completion records are processed and the objects magically appear.
+
+The [staging committer][staging] (aka "Netflix committer") is an older
+committer for S3A.  It requires a "staging" filesystem on HDFS and offers
+manual modes for overwriting outputs ("conflict resolution").
+
+These issues remain with the magic and staging committer:
+
+* Not atomic.  Multipart objects appear sequentially.  Finding a "_success"
+  object indicator is still required before a directory tree can be
+  processed.
+* The magic committer requires "magic committer" support in the FileSystem,
+  in order to write `__magic` objects to a _different_ path.  (Current
+  lakeFSFS does not contain this support.)
+* The magic committer is somewhat new:
+  > It’s also not been field tested to the extent of Netflix’s committer;
+  > consider it the least mature of the committers.
+* Documentation is still somewhat lacking.
+
+Other OutputCommitters are not as good, and _rename by copying_.  This is
+both slow and highly non-atomic, as well as making it difficult to recover
+from failed attempts.  The current recommendations are to use the
+partitioned staging committer for overwriting or updating partitioned data
+trees, and the magic committer is most other cases.
+
+## LakeFSOutputCommitter
+
+We propose to leverage the atomic capabilities of lakeFS to write a specific
+OutputCommitter for lakeFSFS.  In the initial version, it will branch out,
+prepare the desired output, and merge back in as part of the HOC commit.[^1]
+Aborting will be done by dropping the branch (or repurposing it if the same
+job ID is requested again).  Regular lakeFS retention can handle dropping
+data objects; we might want to add file patterns to retention to allow
+temporary objects to be dropped rapidly.
+
+[^1]: A merge is a type of commit, of course, so in this model HOC commits
+	*are* lakeFS commits!
+
+# How
+
+## Conflict modes
+
+Spark supports multiple "save modes": Append, Overwrite, ErrorIfExists, and
+Ignore.  These impact conflict resolution.  We will initially support just
+overwrite modes: an entire previous table will be deleted on write.
+
+## Sample flow
+
+* User configures lakeFSFS and configures LakeFSOutputCommitter as the
+  OutputCommitter for `lakefs` protocol Paths.  **Possibly** lakeFSFS will
+  set up this OutputCommitter by default.
+
+  Successive steps are controlled by Spark/Hadoop to output, and correspond
+  to the Hadoop OutputCommitter protocol.
+* **Setup (job/task TBD)**: Create a new branch for this job/task.  Its name
+  is predictable from the job ID and/or task ID, so can be easily found
```

Comment:
Do we need to easily find it for troubleshooting purposes?

Reply:
Yes, or for debugging, or for cleanup if things go wrong.